src/dsp_tools/utils/xmlupload/xmlupload.py

```diff
@@ -118,15 +118,16 @@ def _prepare_upload(
     shortcode: str,
     verbose: bool,
 ) -> tuple[list[XMLResource], dict[str, Permissions], dict[str, type], Stash | None]:
+    stash_lookup, upload_order = identify_circular_references(root, verbose=verbose)
     resources, permissions_lookup, resclass_name_2_type = _get_data_from_xml(
         con=con,
         root=root,
         default_ontology=default_ontology,
         shortcode=shortcode,
         verbose=verbose,
     )
-    # temporarily remove circular references
-    resources, stash = remove_circular_references(resources, verbose=verbose)
+    resources = sorted(resources, key=lambda res: upload_order.index(res.id))
```

Comment:
I'm a bit concerned about the performance of this: the way I understand sorting in python, the lambda is called for each element in the list; and looking up the index of a list is O(n). So in the worst case we end up with O(n^2) here.  
If I understand the code correctly, we could get this down to O(n), if resources was a dict {res.id: res}, and then we could say `sorted_resources = [resources[res_id] for res_id in upload_order]`.

I'm happy with either of the following two options:
1. you merge this as is, and then check if it's a performance issue with large datasets.
2. you try to address the potential issue in this PR.

But I wouldn't like to take our chances and just hoping that this doesn't slow us down.

Reply:
Wow, it was indeed a huge performance problem: With SGV, my sorting approach took 4 minutes, yours takes 0.1 seconds... Thanks for pointing this out!

Reply:
cool, I didn't even expect that much difference